### PR TITLE
LTSS registration in HanaSR test

### PR DESCRIPTION
**Pull request by mpagot**
- Related ticket: https://jira.suse.com/browse/TEAM-9839

# Verification run:

## BYOS
No LTSS ADDON sle-12-SP5-HanaSr-Azure-Byos-x86_64-Build12-SP5_2024-11-12T03:03:21Z-hanasr_azure_test_sbd az_Standard_E4s_v3
- http://openqaworker15.qa.suse.cz/tests/302892 :green_circle: 

No LTSS ADDON  and also including https://github.com/SUSE/qe-sap-deployment/pull/295. sle-12-SP5-HanaSr-Azure-Byos-x86_64-Build12-SP5_2024-11-12T03:03:21Z-hanasr_azure_test_sbd az_Standard_E4s_v3
- http://openqaworker15.qa.suse.cz/tests/302982 :green_circle: 

SCC_ADDONS=ltss but no regcode  (expected to fails) sle-12-SP5-HanaSr-Azure-Byos-x86_64-Build12-SP5_2024-11-12T03:03:21Z-hanasr_azure_test_sbd az_Standard_E4s_v3
- http://openqaworker15.qa.suse.cz/tests/302992 :green_circle: 
- http://openqaworker15.qa.suse.cz/tests/303000

And this is with intentionally wrong code (expected to fails)
-  http://openqaworker15.qa.suse.cz/tests/302991 :green_circle: 

SCC_ADDONS=ltss  and SCC_REGCODE_LTSS sle-12-SP5-HanaSr-Azure-Byos-x86_64-Build12-SP5_2024-11-12T03:03:21Z-hanasr_azure_test_sbd az_Standard_E4s_v3
- http://openqaworker15.qa.suse.cz/tests/302980

BYOS **with ltss** and so test code should enable use_suseconnect. Also including https://github.com/SUSE/qe-sap-deployment/pull/295. sle-12-SP5-HanaSr-Azure-Byos-x86_64-Build12-SP5_2024-11-12T03:03:21Z-hanasr_azure_test_sbd az_Standard_E4s_v3
-  http://openqaworker15.qa.suse.cz/tests/302981 :green_circle: 
-  http://openqaworker15.qa.suse.cz/tests/302989
-  http://openqaworker15.qa.suse.cz/tests/302990

## PAYG
Has BYOS in the openQA name but it is PAYG, without ltss. Also including https://github.com/SUSE/qe-sap-deployment/pull/295

sle-12-SP5-HanaSr-Azure-Byos-x86_64-Build12-SP5_2024-11-12T03:03:21Z-hanasr_azure_test_sbd az_Standard_E4s_v3
- http://openqaworker15.qa.suse.cz/tests/302983

**with ltss**  Also including https://github.com/SUSE/qe-sap-deployment/pull/295
sle-12-SP5-HanaSr-Azure-Byos-x86_64-Build12-SP5_2024-11-12T03:03:21Z-hanasr_azure_test_sbd az_Standard_E4s_v3
- http://openqaworker15.qa.suse.cz/tests/302984

### CODE REVIEW

File: lib/sles4sap_publiccloud.pm [hunk 1]
```diff
@@ -921,14 +928,18 @@ sub create_playbook_section_list {
...
 
     unless ($args{registration} eq 'noreg') {
-        # Add registration module as first element
-        my $reg_code = '-e reg_code=' . get_required_var('SCC_REGCODE_SLES4SAP') . " -e email_address=''";
-        if ($args{registration} eq 'suseconnect') {
-            push @playbook_list, "registration.yaml $reg_code -e use_suseconnect=true";
-        }
-        else {
-            push @playbook_list, "registration.yaml $reg_code";
+        my @reg_args = ('registration.yaml');
```
Comment (mpagot) [inline lib/sles4sap_publiccloud.pm:931]: as the command line composition for this playbook become more complex I decided to fit all argument in a list and then join them at the end

File: lib/sles4sap_publiccloud.pm [hunk 2]
```diff
@@ -921,14 +928,18 @@ sub create_playbook_section_list {
...
+        push @reg_args, '-e reg_code=' . get_required_var('SCC_REGCODE_SLES4SAP') . " -e email_address=''";
+        push @reg_args, '-e use_suseconnect=true' if ($args{registration} eq 'suseconnect');
```
Comment (mpagot) [inline lib/sles4sap_publiccloud.pm:933]: changed to inline

File: lib/sles4sap_publiccloud.pm [hunk 3]
```diff
@@ -921,14 +928,18 @@ sub create_playbook_section_list {
...
+        if ($args{ltss}) {
+            my @ltss_args = split(/,/, $args{ltss});
```
Comment (mpagot) [inline lib/sles4sap_publiccloud.pm:935]: unpack the argument, symmetrical to pack in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20642/files#diff-d537e7676e4de934957a302dd9662d07237c473a02581e3fa2ca4065a6471c6dR190

File: tests/sles4sap/publiccloud/qesap_terraform.pm [hunk 4]
```diff
@@ -175,6 +176,21 @@ sub run {
             $playbook_configs{spn_application_password} = get_var('AZURE_SPN_APP_PASSWORD', get_required_var('_SECRET_AZURE_SPN_APP_PASSWORD'));
         }
     }
+
+    my @addons = grep { defined $_ && $_ } split(/,/, get_var('SCC_ADDONS'));
+    # This implementation has a known limitation
+    # if SCC_ADDONS has two or more elements (like "ltss,ltss_es")
+    # only the last one will be added to the playbook argument.
+    foreach my $addon (@addons) {
```
Comment (mpagot) [inline tests/sles4sap/publiccloud/qesap_terraform.pm:184]: Scan all extensions in SCC_ADDONS and only care about ltss, calculate SCC Extension name and reg code for it and pack them in the right format expected by ltss argument of create_playbook_section_list

File: tests/sles4sap/publiccloud/qesap_terraform.pm [hunk 5]
```diff
@@ -175,6 +176,21 @@ sub run {
...
+        my $name;
+        # Keep the code simple by only support ltss addons,
+        # it simplify version calculation.
+        $name = get_addon_fullname($addon) if ($addon =~ 'ltss');
```
Comment (alvarocarvajald) [inline tests/sles4sap/publiccloud/qesap_terraform.pm:188]: What happens when `$addon` is not LTSS?

File: tests/sles4sap/publiccloud/qesap_terraform.pm [hunk 6]
```diff
@@ -175,6 +176,21 @@ sub run {
...
+        if (length $name) {
```
Comment (alvarocarvajald) [inline tests/sles4sap/publiccloud/qesap_terraform.pm:189]: `if ($name)` should be good enough.

File: tests/sles4sap/publiccloud/qesap_terraform.pm (hunk 5 above)
Comment (mpagot) [reply tests/sles4sap/publiccloud/qesap_terraform.pm:188]: nothing, no extension added to the playbook argument

File: tests/sles4sap/publiccloud/qesap_terraform.pm (hunk 6 above)
Comment (mpagot) [reply tests/sles4sap/publiccloud/qesap_terraform.pm:189]: yes, I get it from https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/602a746062ede11b555eb2f62da4bbed7b3fdead/lib/registration.pm#L322

**Review by alvarocarvajald — APPROVED** [review]
LGTM

File: lib/sles4sap_publiccloud.pm [hunk 7]
```diff
@@ -17,23 +17,22 @@ use strict;
...
 use utils qw(file_content_replace);
 use serial_terminal qw(serial_term_prompt);
-use version_utils qw(check_version);
+use version_utils qw(check_version is_sle);
 use hacluster;
 use qesapdeployment;
 use publiccloud::utils;
 use publiccloud::provider;
-use publiccloud::ssh_interactive 'select_host_console';
+use publiccloud::ssh_interactive qw(select_host_console);
```
Comment (lpalovsky) [inline lib/sles4sap_publiccloud.pm:32]: :+1:

**Review by lpalovsky — APPROVED** [review]
LGTM

**Pull request merged**